src/Dialect/ONNX/ElementsAttr/ElementsAttrBuilder.cpp

```diff
@@ -450,6 +450,153 @@ std::vector<ElementsAttr> ElementsAttrBuilder::split(
   return results;
 }
 
+ElementsAttr ElementsAttrBuilder::concat(
+    ArrayRef<ElementsAttr> elms, unsigned axis) {
+  assert(elms.size() >= 1 && "concat tensors must be non-empty");
+
+  ElementsAttr first = elms.front();
+  auto firstShape = first.getType().getShape();
+  size_t rank = firstShape.size();
+  assert(axis < rank && "concat axis out of range");
+  if (elms.size() == 1)
+    return first;
+
+  // Check elms are compatible and construct outShape:
+  SmallVector<int64_t> outShape(firstShape);
+  for (size_t i = 1; i < elms.size(); ++i) {
+    ElementsAttr next = elms[i];
+    assert(next.getElementType() == first.getElementType() &&
+           "concat tensors element types must agree");
+    auto nextShape = next.getType().getShape();
+    assert(nextShape.size() == rank && "concat tensors ranks must agree");
+    for (unsigned a = 0; a < rank; ++a) {
+      assert((a == axis || nextShape[a] == firstShape[a]) &&
+             "concat tensors shapes must agree except for the concat axis");
+    }
+    outShape[axis] += nextShape[axis];
+  }
+
+  ShapedType outType = first.getType().clone(outShape);
+  return fromWideNums(outType, [&](MutableArrayRef<WideNum> dst) {
+    auto postAxisShape = ArrayRef(outShape).drop_front(axis + 1);
+    size_t postAxisNumElements = ShapedType::getNumElements(postAxisShape);
+    // A "block" fixes the axes before axis and iterates over the others.
+    size_t outBlockLen = outShape[axis] * postAxisNumElements;
+    size_t start = 0;
+    for (ElementsAttr input : elms) {
+      ArrayRef<int64_t> inputShape = input.getType().getShape();
+      size_t inputBlockLen = inputShape[axis] * postAxisNumElements;
+      SmallVector<int64_t> strides;
+      ArrayBuffer<WideNum> src = getWideNumsAndStrides(input, strides);
+      StridesRange<1> range(inputShape, {strides});
+      auto it = range.begin();
+      for (size_t offset = start; offset < dst.size(); offset += outBlockLen) {
+        for (size_t pos = 0; pos < inputBlockLen; ++pos) {
+          dst[offset + pos] = src.get()[it->pos[0]];
+          ++it;
+        }
```

Comment:
Is it possible to do a memcpy for a block?

Reply:
> that won't work in the cases where the values in input are not contiguous, namely when input is splat or a DisposableElementsAttr with non-default strides

Can we check `splat` or `non-default` strides? It looks like your DisposableElementsAttr can do it.

Reply:
In case of `splat`, we can do `memset` starting from `dst[offset]`

Reply:
> Can we check splat or non-default strides? It looks like your DisposableElementsAttr can do it.

yes, we can test whether the ElementsAttr is splat or whether the strides describe a contiguous row-major layout, but it adds more conditional logic and it will require a lot of tests to get good test coverage

it it ok if I try to add these optimizations in a follow up PR? I can create an issue to track it

Reply:
Sure, it's totally fine.

Reply:
I created issue #2197

Reply:
Thank you!